File: apollo-runtime/src/main/java/com/apollographql/apollo/cache/normalized/lru/EvictionPolicy.java
package com.apollographql.apollo.cache.normalized.lru;

import com.apollographql.android.api.graphql.internal.Optional;

import java.util.concurrent.TimeUnit;

public final class EvictionPolicy {

  public static final EvictionPolicy NO_EVICTION = EvictionPolicy.builder().build();

  private final Optional<Long> maxSizeBytes;
  private final Optional<Long> maxEntries;
  private final Optional<Long> expireAfterAccess;
  private final Optional<TimeUnit> expireAfterAccessTimeUnit;
  private final Optional<Long> expireAfterWrite;
  private final Optional<TimeUnit> expireAfterWriteTimeUnit;

  Optional<Long> maxSizeBytes() {
    return maxSizeBytes;
  }

  Optional<Long> maxEntries() {
    return maxEntries;
  }

  Optional<Long> expireAfterAccess() {
    return expireAfterAccess;
  }

  Optional<TimeUnit> expireAfterAccessTimeUnit() {
    return expireAfterAccessTimeUnit;
  }

  Optional<Long> expireAfterWrite() {
    return expireAfterWrite;
  }

  Optional<TimeUnit> expireAfterWriteTimeUnit() {
    return expireAfterWriteTimeUnit;
  }

  public static EvictionPolicy.Builder builder() {
    return new EvictionPolicy.Builder();
  }

  public static class Builder {

    private Builder() { }

    private Optional<Long> maxSizeBytes = Optional.absent();
    private Optional<Long> maxEntries = Optional.absent();
    private Optional<Long> expireAfterAccess = Optional.absent();
    private Optional<TimeUnit> expireAfterAccessTimeUnit = Optional.absent();
    private Optional<Long> expireAfterWrite = Optional.absent();
    private Optional<TimeUnit> expireAfterWriteTimeUnit = Optional.absent();

    public EvictionPolicy.Builder maxSizeBytes(long maxSizeBytes) {
      this.maxSizeBytes = Optional.of(maxSizeBytes);
      return this;
    }

    public EvictionPolicy.Builder maxEntries(long maxEntries) {
      this.maxEntries = Optional.of(maxEntries);
      return this;
    }

    public EvictionPolicy.Builder expireAfterAccess(long time, TimeUnit timeUnit) {
      this.expireAfterAccess = Optional.of(time);
      this.expireAfterAccessTimeUnit = Optional.of(timeUnit);
      return this;
    }

    public EvictionPolicy.Builder expireAfterWrite(long time, TimeUnit timeUnit) {
      this.expireAfterWrite = Optional.of(time);
      this.expireAfterWriteTimeUnit = Optional.of(timeUnit);
      return this;
    }

    public EvictionPolicy build() {
      return new EvictionPolicy(maxSizeBytes, maxEntries, expireAfterAccess, expireAfterAccessTimeUnit,
          expireAfterWrite, expireAfterWriteTimeUnit);
    }

  }

  private EvictionPolicy(Optional<Long> maxSizeBytes, Optional<Long> maxEntries, Optional<Long> expireAfterAccess,
      Optional<TimeUnit> expireAfterAccessTimeUnit, Optional<Long> expireAfterWrite, Optional<TimeUnit>
      expireAfterWriteTimeUnit) {
    this.maxSizeBytes = maxSizeBytes;
    this.maxEntries = maxEntries;
    this.expireAfterAccess = expireAfterAccess;
    this.expireAfterAccessTimeUnit = expireAfterAccessTimeUnit;
    this.expireAfterWrite = expireAfterWrite;
    this.expireAfterWriteTimeUnit = expireAfterWriteTimeUnit;
  }

}

File: apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/lru/EvictionPolicy.kt
package com.apollographql.apollo.cache.normalized.lru

import com.apollographql.apollo.api.internal.Optional
import java.util.concurrent.TimeUnit

/**
 * Controls how long a [com.apollographql.apollo.cache.normalized.Record] will
 * stay in a [LruNormalizedCache].
 */
class EvictionPolicy internal constructor(
    private val maxSizeBytes: Optional<Long>,
    private val maxEntries: Optional<Long>,
    private val expireAfterAccess: Optional<Long>,
    private val expireAfterAccessTimeUnit: Optional<TimeUnit>,
    private val expireAfterWrite: Optional<Long>,
    private val expireAfterWriteTimeUnit: Optional<TimeUnit>
) {

  fun maxSizeBytes(): Optional<Long> {
    return maxSizeBytes
  }

  fun maxEntries(): Optional<Long> {
    return maxEntries
  }

  fun expireAfterAccess(): Optional<Long> {
    return expireAfterAccess
  }

  fun expireAfterAccessTimeUnit(): Optional<TimeUnit> {
    return expireAfterAccessTimeUnit
  }

  fun expireAfterWrite(): Optional<Long> {
    return expireAfterWrite
  }

  fun expireAfterWriteTimeUnit(): Optional<TimeUnit> {
    return expireAfterWriteTimeUnit
  }

  class Builder internal constructor() {
    private var maxSizeBytes = Optional.absent<Long>()
    private var maxEntries = Optional.absent<Long>()
    private var expireAfterAccess = Optional.absent<Long>()
    private var expireAfterAccessTimeUnit = Optional.absent<TimeUnit>()
    private var expireAfterWrite = Optional.absent<Long>()
    private var expireAfterWriteTimeUnit = Optional.absent<TimeUnit>()

    fun maxSizeBytes(maxSizeBytes: Long): Builder {
      this.maxSizeBytes = Optional.of(maxSizeBytes)
      return this
    }

    fun maxEntries(maxEntries: Long): Builder {
      this.maxEntries = Optional.of(maxEntries)
      return this
    }

    fun expireAfterAccess(time: Long, timeUnit: TimeUnit): Builder {
      expireAfterAccess = Optional.of(time)
      expireAfterAccessTimeUnit = Optional.of(timeUnit)
      return this
    }

    fun expireAfterWrite(time: Long, timeUnit: TimeUnit): Builder {
      expireAfterWrite = Optional.of(time)
      expireAfterWriteTimeUnit = Optional.of(timeUnit)
      return this
    }

    fun build(): EvictionPolicy {
      return EvictionPolicy(
          maxSizeBytes,
          maxEntries,
          expireAfterAccess,
          expireAfterAccessTimeUnit,
          expireAfterWrite,
          expireAfterWriteTimeUnit
      )
    }
  }

  companion object {
    val NO_EVICTION = builder().build()
    fun builder(): Builder {
      return Builder()
    }
  }

}